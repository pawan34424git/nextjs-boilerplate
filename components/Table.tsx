import { getElKey } from "@/utils/helper-utils";
import TableRow from "./TableRow";

export interface ITableColumnData {
  value: string | number;
  header?: string;
  cb?: (data: ITableColumnData) => any;
}

export interface ITableRowData {
  [key: string]: ITableColumnData;
}
const DUMMY_DATA: ITableRowData[] = [
  {
    col1: {
      header: "Col1",
      value: "test",
    },
    col2: {
      header: "Co2",
      value: "fdgd",
    },
    action: {
      value: "edit",
      cb: () => {
        console.log("clicked Edit");
      },
    },
  },
];

interface ITable {
  data?: ITableRowData[];
}
const Table = ({ data = DUMMY_DATA }: ITable) => {
  const cols = Object.keys(data[0]);
  const displayHeaders = cols.map((key) => data[0][key].header || key);

  return (
    <>
      <div className="shadow-2x rounded-2xl bg-white p-2 min-w-full">
        <table className="table-auto min-w-full">
          <thead>
            <tr>
              {displayHeaders.map((label) => (
                <th key={label} className="px-4 py-2 border">
                  {label}
                </th>
              ))}
            </tr>
          </thead>
          <tbody>
            {data.map((row, index) => (
              <TableRow cols={cols} row={row} key={getElKey(row, index)} />
            ))}
          </tbody>
        </table>
      </div>
    </>
  );
};

export default Table;
