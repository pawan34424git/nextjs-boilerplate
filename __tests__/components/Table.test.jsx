import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import Table from "../../components/Table";

describe("Table Component", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Table />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("show table with default value", () => {
    render(<Table />);
    expect(screen.getByTestId("table")).toBeInTheDocument();
    expect(screen.queryByTestId("table")).toBeTruthy();

    const editButton = screen.getByText("edit");
    fireEvent.click(editButton);
  });

  it("will not show table", () => {
    render(<Table data={[]} />);
    expect(screen.queryByTestId("table")).toBeFalsy();
  });
});
