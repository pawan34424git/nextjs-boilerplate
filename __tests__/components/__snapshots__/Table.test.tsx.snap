// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component table rendering 1`] = `
<div
  className="shadow-2x rounded-2xl bg-white p-2 min-w-full"
  data-testid="table"
>
  <table
    className="table-auto min-w-full"
  >
    <thead>
      <tr>
        <th
          className="px-4 py-2 border"
        >
          Col1
        </th>
        <th
          className="px-4 py-2 border"
        >
          Co2
        </th>
        <th
          className="px-4 py-2 border"
        >
          action
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="border px-4 py-2  font-medium"
        >
          test
        </td>
        <td
          className="border px-4 py-2  font-medium"
        >
          fdgd
        </td>
        <td
          className="border container py-2 px-2 mx-0 min-w-full flex flex-colalign-top justify-center items-center"
        >
          <button
            className="bg-purple-900 text-white hover:text-seconday hover:bg-primary font-bold py-2 px-4 rounded"
            onClick={[Function]}
          >
            edit
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
